### DISCUSSION TITLE
Best practices for order of ops

### DISCUSSION THREAD

**Discussion by mediajolt001** [cables-gl/cables_docs#626] [Answers & Questions] [answered]
Is there any "best practices" documentation explaining the order in which ops should be connected? This is the single hardest thing for me to parse in learning Cables - when do I add a renderer, when do I add a mesh? I know I can follow the examples, but understanding the reasoning behind the ordering of ops is still a mystery. For example, why does the renderer come BEFORE I've added the items to be rendered? Shouldn't that come afterwards? I'd love some kind of doc that breaks down the order and why - it would make the learning curve much more navigable. IMHO, anyway.

**Comment by pandrr** [comment, marked as answer]
hi,

what do you mean by render ? mainloop was called renderer in the past, which was misleading, maybe you saw that ?

in general the order of the patch is the order of execution. 
yellow lines mean triggering/executing functions.

mainloop triggers the execution of the first operator connected to it (every frame). 

ops will change states, e.g. transforming cameras or positions, setting a shader, in the end everything gets rendered, e.g. by a mesh operator. this will use the current shader(material) and the current transformation(world position and camera etc) and will render itself.

it does not matter if material is set after or before transformation, but they have to be before finaly rendering them to the screen.
